### DISCUSSION TITLE
Proposal: Simplified "If" statement for comparing the same value

### DISCUSSION THREAD

**Discussion by doktrova** [dotnet/csharplang#912] [General]
Lets say we have an if statement like below:

```
if ( Car.Make("Benz").Body.Wheels.Count == 4 || Car.Make("Benz").Body.Wheels.Count == 3)
    //Do something...
```

Since some times the properties chain can be long  and we can have even more ors(||) added. it can get ugly. One solution is to introduce a variable.

```
var numWheels = Car.Make("Benz").Body.Wheels.Count;
if ( numWheels == 4 || numWheels == 3)
  //Do Something..
```

Can we do something where in these cases we don't have to introduce a variable or have long ifs?
Here is an idea. Maybe someone else can have a better looking syntax.
```
if ( Car.Make("Benz").Body.Wheels.Count == 4 || == 3)
    //Do something...

if ( Car.Make("Benz").Body.Wheels.Count == 4 || == 3 && != 26 && >1)
    //Do something...
```

**Comment by mattwar** [comment]
You can already write:

```cs
if ( (Car.Make("Benz").Body.Wheels.Count is var wheels) && (wheels == 4 || wheels == 3) )
   // Do something...
```

**Comment by bondsbw** [comment]
We had a discussion about a similar feature (partial infix expressions) in #812.

**Comment by TonyValenti** [comment]
I just recently asked for something similar and learned a nice way to do it:
````
if ( Car.Make("Benz").Body.Wheels.Count is var Wheels && (Wheels == 3 || Wheels == 4)) {

}
````

I want to point out that "is var" is not the same as "is <T>".  IsVar always returns true and always assigns a value to the following variable.  Is<T> only returns true if it can assign a non-null value.

**Comment by TonyValenti** [comment]
https://github.com/dotnet/csharplang/issues/899

**Comment by doktrova** [comment]
@mattwar that's pretty cool I didn't know about that, but In my opinion it is the same amount of code and it looks more confusing than declaring a new variable.

**Comment by DavidArno** [comment]
This can be worked around using `is var`:
```cs
if (Car.Make("Benz").Body.Wheels.Count is var wheelsCount && wheelsCount == 4 || wheelsCount == 3)
```
but it's long-winded. [Allowing `as var`](https://github.com/dotnet/csharplang/issues/402) would make the code tidier:
```cs
if (Car.Make("Benz").Body.Wheels.Count as var wheelsCount == 4 || wheelsCount == 3)
```

**Comment by Pzixel** [comment]
Am I alone thinkinh that using `is` for just create a variable where value goes is a broken idea? Really, we already have a syntax for declare a variable `var foo = 42`, so I don't understand at all why even write `if (42 is var foo)`. The former is nicier in all ways. I think use `is` for declaring variables is just a feature abuse.

I'd like more idea about expression `if` or blocked `a ? { b } : { c }` operator but that's another story...

**Comment by HaloFour** [comment]
@Pzixel 

It's not `is` that is creating the variable, it is the variable pattern `var`.  The use in these simple cases could be considered "abuse" only in that it's weird shorthand for something someone could already do if they weren't so bent on single-line expressions.  Where variable patterns will become more useful is in recursive pattern matching where it can be used to extract a value from within a more complicated expression.

Either way, it's in the language already.

And you already have an expression `if`, it's called `?:`.

Conversely, you already have a blocked `?:`, it's called `if`.

**Comment by Pzixel** [comment]
@HaloFour when I'm taling about blocking if I'm talking about if's like in Rust
```rust
    let big_n =
        if n < 10 && n > -10 {
            println!("n is a small number, increase ten-fold");
            10 * n
        } else {
            println!("n is a big number, reduce by two");
            n / 2
        };
```

Today I'm trying to extract a method where I can but it's not always possible/preferable.
So extending either `if` to be an expression or `? : ` to accept statements would do it. But being said, that's another story. And it may be solved with upcoming `match` keyword.

> It's not is that is creating the variable, it is the variable pattern var

I was never understanding this `always matching pattern`. Why even introduce it? I can understand `if ((object)42 is int i && i > 10) ..`, but if pattern always succeed why not just store it in a ol' good way? Maybe there are some scenarious, but I haven't seen any.

**Comment by HaloFour** [comment]
@Pzixel 

In Rust all statements are either declarations or expressions.  That is not the case in C# and it wouldn't make sense to try to force it onto C#.

> I was never understanding this `always matching pattern`.  Why even introduce it?

Sets the foundation for recursive patterns where some other part of the pattern might not succeed.  Or it could be combined with pattern guards to filter based on more complex logic that can't be directly mapped to a pattern.

The use case described in this proposal isn't a *wrong* use of the `var` pattern.  But if you don't like it you're free to continue pre-declaring and assigning your variables prior to comparing them.

Note that Rust has top-level always-matching variable bindings also:

```rust
let x = 1;

match x {
    y => println!("x: {} y: {}", x, y),
}
```

**Comment by Pzixel** [comment]
> In Rust all statements are either declarations or expressions. That is not the case in C# and it wouldn't make sense to try to force it onto C#.

Do not substitute concepts, I'm talking about extensing `? :` operator, not about having all blocks return values :)

> The use case described in this proposal isn't a wrong use of the var pattern. But if you don't like it you're free to continue pre-declaring and assigning your variables prior to comparing them

It's just my IMHO that that's a misuse of the feature. No more, no less.

> Sets the foundation for recursive patterns where some other part of the pattern might not succeed. Or it could be combined with pattern guards to filter based on more complex logic that can't be directly mapped to a pattern.

Could you provide an example, please?

**Comment by DavidArno** [comment]
@Pzixel,

One example of recursive patterns would be with tuples:
```cs
(bool succss, int value) TryParse(string s) ... 

if (TryParse("123") is var (success, value) && success)
{
    // use value here
```

**Comment by HaloFour** [comment]
@Pzixel 

> Do not substitute concepts,

You're the one who mentioned expression `if`.

Take a look at #377, that would enable the scenario that you described:

```cs
var big_n = (n < 10 && n > -10)
    ? (Console.WriteLine(", and is a small number, increase ten-fold"); 10 * n)
    : (Console.WriteLine(", and is a big number, reduce by two"); n / 2);
```

**Comment by DavidArno** [comment]
@Pzixel,
> Am I alone thinkinh that using is for just create a variable where value goes is a broken idea?

I doubt you are alone, but I disagree it's a broken idea. I have to confess, I haven't found a use for `x is var y` yet, but I use `x is T y` a lot. It provides a compact way of testing whether a variable is of a given type and introducing a new variable of that type all in one go:
```cs
SomeType Foo(IEnumerable<T> collection) =>
    collection is List<T> list ? HandleAsList(list) : HandleAsEnumerable(collection);
```

**Comment by Pzixel** [comment]
@DavidArno this example is completely valid. I'm talking about some bizzare things like:

```csharp
SomeType Foo() =>
   new int[1,2,3] is var collection ? HandleAsEnumerable(collection) : null;
```

instead of 

```csharp
SomeType Foo()
{
   var collection = new int[1,2,3];
   HandleAsEnumerable(collection);
}
```

just because it's singleline.

Here is plenty example in issues where they are using just to copy a value:

```csharp
SomeType Foo()
{
   var collection = new int[1,2,3];
   if (collection is var collectionCopy)
   {
      HandleAsEnumerable(collectionCopy);
   }
}
```
or more realistic example:
```csharp
if ( (Car.Make("Benz").Body.Wheels.Count is var wheels) && (wheels == 4 || wheels == 3) )
```
and so on.

**Comment by DavidArno** [comment]
@Pzixel,

We used to have to write:
```cs
int x;
if (int.TryParse("123", out x) ...
```
Now we can write it on one line:
```cs
if (int.TryParse("123", out int x) ...
```
Do you see this feature as a broken idea?

Assuming no, why do you see the following as a broken idea:
```cs
if ((Car.Make("Benz").Body.Wheels.Count as var wheels) == 4 || wheels == 3)
```
Why is typing it longhand better?
```cs
var wheels = Car.Make("Benz").Body.Wheels.Count;
if (wheels == 4 || wheels == 3)
```
Currently, we have to do `Car.Make("Benz").Body.Wheels.Count is var wheels && wheels ==`, but again why is this a broken idea when `out var` isn't? There seem to be a group of people here indulging in double standards: they love the idea of introducing the variable for an `out` parameter at the point of use, but oppose the idea of introducing other variables into expressions at the point of use, bemoaning "The only benefit I see is that it lets you cram more onto a single line expression" (as @HaloFour has put it before).

Likewise (taken from #402), would you view this ~~improvement~~ alternative syntax as a bad idea?
```cs
var pointMass1 = new PointMass(mass: 1.0f);
var pointMass2 = new PointMass(mass: 2.0f);
var simulation = new Simulation
{
    pointMass1,
    pointMass2,
    new Spring(pointMass1, pointMass2)
};
```
becomes:
```cs
var simulation = new Simulation
{
    new PointMass(mass: 1.0f) as var pointMass1,
    new PointMass(mass: 1.0f) as var pointMass2,
    new Spring(pointMass1, pointMass2)
};
```
Limiting the scope of variables is a good thing (unless you are a fan of "globals for the win"). Ergo, if a variable is only needed in an expression, define it in the expression. This seems common sense to me.

**Comment by orthoxerox** [comment]
I would rather use sequence expressions for this than `as var`, the latter doesn't look legible at all to me.

```c#
if ((
    var wheels = Car.Make("Benz").Body.Wheels.Count;
    wheels == 4 || wheels == 3)) { ...
```

Using it in the object initializer example doesn't produce a similarly nice result, though, but is still legible.

```c#
var simulation = (
    var pointMass1 = new PointMass(mass: 1.0f);
    var pointMass2 = new PointMass(mass: 1.0f);
    new Simulation
    {
        pointMass1,
        pointMass2,
        new Spring(pointMass1, pointMass2)
    });
```

**Comment by Pzixel** [comment]
> Do you see this feature as a broken idea?
No, because it's some sugar for a boilerplate variable declaration that we had to write until now.

> Assuming no, why do you see the following as a broken idea:

Yes, because we already have the nice syntax variable declaration. 

> Why is typing it longhand better?

Because it express your intent while the latter disguise "i'm not an assgnment, i'm a boolean expression" which is actually not.

> Likewise (taken from #402), would you view this improvement as a bad idea?

You always used `improvement` word so if I say `no` I look like a retrograde :) So I can't 

But seriosly, this feature allow you to express better your intent "this variables belong to this scope only", so it's a good thing. However it has nothing common with this "look, you think i'm a type check operator, but hehe, you're wrong".

**Comment by DavidArno** [comment]
@Pzixel,

I have changed "improvement" (which I clearly see it as) to "alternative syntax" so as to not prejudice your answer.

So you are saying you like `out var` because it lets you replace "nice syntax variable declaration" with a boolean expression that hides an assignment, but you dislike `x is var y` because "it's some sugar for a boilerplate variable declaration that we had to write until now".

See what I mean about double standards?

**Comment by Pzixel** [comment]
No, I don't.
```
IDictionary<string, object> dictionary;
if (myparser.TryParse(input, out dictionary)
   ...
```
=>
```
if (myparser.TryParse(input, out var dictionary)
   ...
```
it's an obvious improvement.

`int i = 42;` => `42 is var i` - yoda style assignment that gives you nothing but yet another way to express your intent that already has special syntax, but in much more confusing manner.

Thus I don't see any double standards. I always accept another opition when I'm convinced that I'm not right, because it's dangerous to be ignorant to faults. But it's not the case. These features are completely different. One adds obvious value, second adds yet another ways to write existing things in more confusing manner, imho. I'm completely fine with `a is T b` because it's just a shortcut for `var b = a as T; if  b != null`, with little magic for value types. It's wonderful, I like this feature. But match that always succeed is nothing but an assignment.

**Comment by HaloFour** [comment]
@Pzixel 

You're free to not use it.  It's already in the language so arguing over it is kind of moot.  Maybe you'll understand it's value proposition better when recursive patterns roll out.

**Comment by bondsbw** [comment]
`if (42 is var i) {...}` is a red herring, sort of like complaining that `while (false) {...}` is legal syntax.  I don't know that anyone actually uses it that way.

**Comment by Pzixel** [comment]
@HaloFour all I do is trying to understand this value. This feature came to language so probably lots of smart people proposed and developed thus it **must** have some value. But unfortunetly, I don't see it yet. All this discussion is about this understanding.

@bondsbw nope. `if (42 is var i) {...}` is just `if (Car.Make("Benz").Body.Wheels.Count is var wheels)`, which has been used here and many similar statements across all the repository. You can't say `nobody write this way` when they actually are.

**Comment by DavidArno** [comment]
@Pzixel,
```cs
if (42 is var i) {...}
```
is a silly construct, that we all agree would be a bad way to write code. You are using it to ridicule sensible uses of the `x is var y` construct. It's a strawman. Attacking a strawman, rather than the real thing, is a classic logical fallacy.

Whereas,
```cs
if (Car.Make("Benz").Body.Wheels.Count is var wheels && wheels == 4 || wheels == 3) {...}
```
is viewed as a sensible use of `x is var y`. If you disagree, then I'd suggest you stick to using the latter in your arguments as it adds more weight to those arguments. Though it's of course harder to attack real uses rather than a strawman...

**Comment by Pzixel** [comment]
@DavidArno it's called _example_, to focus on actual problem instead of details. But if it confuses anyone then ok.
How
`if (Car.Make("Benz").Body.Wheels.Count is var wheels && wheels == 4 || wheels == 3) {...}`
is better then
```csharp
var wheels = Car.Make("Benz").Body.Wheels.Count;
if (wheels == 4 || wheels == 3) {...}
```
?

We also might see that the former is `if (a && b || c)` not `if (a && (b || c))` but `a` is always true then ok. However it's also confusing, again imho.

I just want to understand why you think that the former is better then the latter. Because my logic is simply `2 is easier to write than 1, but it's harder to read. Readability > writeabiliy, thus 2 is better`. What logic makes you think that 1 is better? 

Why I think that 1 is readable? Because void statement that assign a value is better, then an expression that always return true. Statement `a` is always better than `if (true) { a }`.

This logic is simple and I don't where it may be wrong. But you have another opinion so it probably has some.

**Comment by DavidArno** [comment]
In my view, it's better because it is making it clear that the scope of `wheels` is confined to the `if/else` clause. Of course, there is a caveat to that: `wheels` leaks out into the surrounding scope due to (in my opinion at least) a poor design decision by the language team. But that is easily "fixed" with an analyzer that reports the use of that leaked variable outside that scope as a warning or error.

**Comment by jnm2** [comment]
@DavidArno You keep saying that, and I keep using it for early returns or continues and for throwing exceptions and liking it.

![image](https://user-images.githubusercontent.com/8040367/30650104-0e5735f2-9df0-11e7-9594-39d7025e420f.png)

**Comment by DavidArno** [comment]
@jnm2,

Could you provide a more detailed example, as I'm not sure I follow you from that snippet?

**Comment by HaloFour** [comment]
@DavidArno 

I believe he means that he's using the "leaky" scope to write validation code where he intends the variable to remain usable:

```cs
if (!(arg is Foo foo))
    throw new ArgumentException($"{nameof(arg)} ain't no Foo!");

foo.Pity();
```

**Comment by jnm2** [comment]
@DavidArno @HaloFour yes, particularly when there are many early exit conditions or when `foo` is referenced in multiple statements following the early exit check.

Also, I dislike leaving the last line of a method a `throw` statement to handle some error condition. I want that to be the main return. The only exception to this is a pure switch or dispatch-style method.

**Comment by jnm2** [comment]
See, corefx does this even more than me:

https://github.com/dotnet/corefx/blob/c4e03f00fe667aed5a4a79a4b039a8c842fdfcc4/src/System.Drawing.Common/src/System/Drawing/Font.cs#L476-L498

**Comment by jnm2** [comment]
Here's a similar example with more detail, using `out var` scope the same way. I see how you could argue the clarity of it either way, but it just feels clean to me:

![image](https://user-images.githubusercontent.com/8040367/30652094-794d7768-9df5-11e7-920e-ec6763f2e726.png)

**Comment by jnm2** [comment]
Oops, missed that I didn't need to assign `info` anymore. Even closer to its usage now. 😄

**Comment by DavidArno** [comment]
@jnm2 & @HaloFour,

Thanks, makes sense now.

Ask me again in a year and I may have changed my mind. In the meantime, I still think it an ugly pattern and the leakiness is too high a price to pay for convenience. If `topLevel` is going to be used beyond the `if`, I'd want to write it as:
```cs
var topLevel = enumTypeReference as TopLevelTypeReference;
if (topLevel == null) throw(...

// use topLevel here.
```

**Comment by jnm2** [comment]
It ends up feeling like a lot of repeated two-statement boilerplate to me. But who knows, maybe in another year I'll agree with you. =D

**Comment by jcouv** [comment]
An extension method would do the trick: `if ( Car.Make("Benz").Body.Wheels.Count.IsIn(3, 4) )`.
You can have a couple of overloads with different numbers of parameters.

**Comment by Pzixel** [comment]
@jcouv `public static bool IsIn(this int value, params Span<int> values) => values.Contains(value);`

**Comment by calin-darie** [comment]
Read this whole thing hoping for a revelation, but I'm sad to say I'm still with @Pzixel here.
```csharp
    if (Car.Make("Benz").Body.Wheels.Count is var wheels && wheels == 4 || wheels == 3)
```
- [pro] it's saving you one instance of typing `wheels` 
- [con] it's making the boolean expression harder to read, you have to figure out it translates to 
```csharp 
if (true && wheels = 4 || wheels == 3)
```
- [con] it's inlining declaration and usage, assignment and control flow. I've got no double standards here, I prefer writing 
```csharp
var success = int.TryParse("42", out i); 
if (!success) return;
```
and it would have been tree lines instead of two with no pattern matching. I really love that pattern matching enables declaration and initialization to go on one line. I don't care about declaration+initialization+(first use). Same applies to the recursive patterns example: 
```csharp 
var (success, value) = TryParse("123");
if (success) {...
```
I generally love where C# syntax is going, and I'm adopting new features early, but in this case the old syntax is clearer for the same amount of keystrokes.
- [con] most important con of all if you ask me *it doesn't solve this issue!* I want to write` if wheels == 4 || == 3 || (>42 && <50)` and saving me just one initial mention of `wheels `is not much of a consolation

Speaking of the original issue, I always wanted a ternary math-like less-than:
```csharp 
if (18 < age && age < 72)
```
=> 
```csharp 
if (18 < age < 72)
```
with all the variations of course
```
< <=
<= <
> >
> >=
>= >
```
Didn't give it a proper thought though. Hopefully there's no ambiguous syntax resulting from these ternary operators.

**Comment by quinmars** [comment]
Why not use a `switch`?

```csharp
switch (Car.Make("Benz").Body.Wheels.Count)
{
    case 3: case 4:
    //Do something...
    break;
}
```

**Comment by calin-darie** [comment]
If all you need to do is test against a few const values, you could go with a `switch`, or, better yet for brevity, with the `.IsIn(` LINQ extension proposed above. But if you need to test against a range, or set of ranges, or some arbitrarily complex expression, a switch won't help much. Also, you can't treat a switch as a boolean expression to further compose it, not without making a method/local function of it anyway.
A solution could be:
```csharp
...
    //usage
    if (42.Is(InRange(30, 50).Or(InRange(100, 200))));
...
    public static class BoolExpressionExtensions
    {
        public static bool Is<T>(this T value, params Predicate<T>[] predicates) => 
            predicates.All(p => p(value));

        public static Predicate<T> InRange<T>(T rangeMin, T rangeMax)
            where T:IComparable<T> => 
            value => value.CompareTo(rangeMin) >= 0 &&
                     value.CompareTo(rangeMax) <= 0;

        public static Predicate<T> Or<T> (this Predicate<T> predicateA, Predicate<T> predicateB) => 
            v => predicateA(v) || predicateB(v);
    }
```
Is there a library for this? I couldn't find one, and I'm tempted to write one.

**Comment by quinmars** [comment]
@calin-darie, not what you are looking for, but [Fluent Assertions](http://fluentassertions.com/) goes in a similar direction.

**Comment by Acerby** [comment]
With pattern matching you can now just say:
if ( Car.Make("Benz").Body.Wheels.Count is 3 or 4)
{
}

**Reply by DavidArno** [reply]
And it's even possible to say:
```cs
if ( Car.Make("Benz").Body.Wheels.Count is 4 or 3 and not 26 and >1)
    //Do something...
```
though I'm less clear why one would *want* to write that 😃 

I wonder if there's more gems like this hidden away in the discussion backlog that have been effectively implement and could be close.

**Comment by AraHaan** [comment]
I could have sworn one could do this today ever since C# 10:
```cs
if ( Car.Make("Benz").Body.Wheels.Count is 3 or 4 )
    //Do something...
```
